swagger/smart_transfer_protocol_ws.yaml

```diff
@@ -0,0 +1,332 @@
+            stopAt:
+              type: string
+              format: multihash
+              description: Hash of the last metadata block consumer would like to pull (undefined for head)
+
+  dataset/pull/response:
+    publish:
+      operationId: odf.dataset.pull.response
+      summary: Indicates succesfull response to dataset pull request
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullResponse
+        name: DatasetPullResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - newBlocks
+          properties:
+            newBlocks:
+              $ref: '#/components/schemas/ObjectsBatch'
+  
+  dataset/pull/objects/negotiate:
+    subscribe:
+      operationId: odf.dataset.pull.objects.negotiate
+      summary: Requests to negotiate the download protocol for 1 or more object files
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsNegotiate
+        name: DatasetPullObjectsNegotiate
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - objectFiles
+          properties:  
+            objectFiles:
+              type: array
+              description: List of object file references in this bulk
+              items:
+                $ref: '#/components/schemas/ObjectFileReference'
+      
+  dataset/pull/objects/negotiate/response:
+    publish:
+      operationId: odf.dataset.pull.objects.negotiate.response
+      summary: Response on negotiated status of object download operations during pull
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPullObjectsNegotiateResponse
+        name: DatasetPullObjectsNegotiateResponse
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - negotiatedDownloads
+          properties:   
+            negotiatedDownloads:
+              type: array
+              description: Negotiated download operations for requested bulk of object files
+              items:
+                $ref: '#/components/schemas/NegotiatedDownload'
+
+  dataset/push/request:
+    subscribe:
+      operationId: odf.dataset.push.request
+      summary: Initiates dataset push to the ODF server
+      security:
+        - jwt: []
+      message:
+        messageId: odf.dataset.DatasetPushRequest
+        name: DatasetPushRequest
+        contentType: application/json
+        payload:
+          type: object
+          required:
+            - datasetId
+            - newBlocks
+          properties:
+            datasetId:
+              type: string
+              format: did
+              description: Dataset identifier
+            newBlocks:
+              $ref: '#/components/schemas/ObjectsBatch'
```

Comment:
I wonder if `odf.dataset.push.request` operation should be more like a "pre-flight" request rather than the first step of sending objects, e.g.:
- client: I want to push to `head` a block interval `(from, to]`  + potentially some estimates on data size / number of files
- server: OK
- client: Here's metadata
- ...

This would give server a chance to do basic checks like permissions, check if `head` is still on the expected hash etc. before client spends time compressing and sending the blocks.

Reply:
I like the idea, but I also extended it for pulls.

Push:
 - client: I want to push new blocks from _head-I-know_ estimated as ...
 - server: OK (permissions checked, size is acceptable)
 - client: Here's metadata, tarball attached
 - server: OK
 - client: let's talk on uploading these N object files
 - server: here's what I suggest
 - _...client sends the objects via obtained URLs..._
 - client: let's commit new head
 - server: commited new head
 - client: closing connection

Pull:
- client: I want to pull dataset from this node to that node
- server: OK, here is a size estimation
- client: Give me metadata
- server: Here you go, tarball attached
- client: let's talk on downloading these N object files
- server: here's what I suggest:
- _client downloads the objects_ via obtained URLs..._
 - client: closing connection

Note that during push, I restored passing the head that client currently knows, this is supposed to allow early divergence detection.